llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -81,10 +81,19 @@ void emitBundlesInScoreboardDelta(
   }
 }
 
-MachineBasicBlock *getSinglePredecessor(const MachineBasicBlock &MBB) {
-  assert(MBB.pred_size() == 1 && "MBB contains more than 1 predecessor");
-  MachineBasicBlock *SinglePredMBB = *MBB.predecessors().begin();
-  return SinglePredMBB;
+MachineBasicBlock *getLoopPredecessor(const MachineBasicBlock &MBB) {
+  if (MBB.pred_size() == 1) {
+    // if we have only one, it must be the loop
+    return *MBB.predecessors().begin();
+  }
+  // Otherwise, the loop is the fallthrough predecessor by construction
```

Comment:
I'm not sure I understand, are we looking for a predecessor of the loop, or for the block which the loop is the predecessor of? I assume the latter given the comment you wrote. Maybe clarify the function name?

Also, how can we guarantee the loop is the fall-through pred, can't we have a branch if MachineBlockPlacement does a bad job?

Reply:
We don't classify something as a loop if the epilogue is not a dedicated exit, unless that exit is fallthrough from the loop. And that classification runs in the same pass. We are looking for the loop, which is a predecessor of the epilogue, and if there are more, it can only be the fallthrough one